### PR TITLE
Enforce Password Generator Policy

### PR DESCRIPTION
**Pull request by vincentsalucci**
## Code Changes
- **abstractions/passwordGeneration.service.ts**: Adjusted abstraction to mirror function calls in `passwordGeneration.service.ts`
- **password-generator.component.ts**: Utilized new `getOptions` function that injects intermediate step for password options retrieval. It applies the rules set by the governing organization to the user's password generation options. UI adjusted by using information stored in the `PasswordGeneratorPolicyOptions` object. Because field's floors were being altered, the `normalizeOptions` function was used to further adjust potential changes.
- **passwordGeneratorPolicyOptions.ts**: Created domain object for storage/retrieval of compiled password generator organization policies
- **passwordGeneration.service.ts**: Adjust `getOptions` function to utilize new `getPasswordGeneratorPolicyOptions` helper method and manipulate the user's options based on the organizations' enabled policy.

### CODE REVIEW

File: src/angular/components/password-generator.component.ts [hunk 1]
```diff
@@ -8,9 +8,7 @@ import {
 import { I18nService } from '../../abstractions/i18n.service';
 import { PasswordGenerationService } from '../../abstractions/passwordGeneration.service';
 import { PlatformUtilsService } from '../../abstractions/platformUtils.service';
-import { PolicyService } from '../../abstractions/policy.service';
-import { PolicyType } from '../../enums/policyType';
-import { Policy } from '../../models/domain/policy';
+import { PasswordGeneratorPolicyOptions } from '../../models/domain/passwordGeneratorPolicyOptions';
```
Comment (kspearrin) [inline src/angular/components/password-generator.component.ts]:
```suggestion

import { PasswordGeneratorPolicyOptions } from '../../models/domain/passwordGeneratorPolicyOptions';
```

File: src/services/passwordGeneration.service.ts [hunk 2]
```diff
@@ -217,7 +223,82 @@ export class PasswordGenerationService implements PasswordGenerationServiceAbstr
             }
         }
 
-        return this.optionsCache;
+        const enforcedPolicyOptions = await this.getPasswordGeneratorPolicyOptions();
+        if (this.optionsCache.length < enforcedPolicyOptions.minLength) {
```
Comment (kspearrin) [inline src/services/passwordGeneration.service.ts]: Is there a way you can just return `null` here if there are no policies to enforce? Then you could just skip all of this condition checking.

File: src/services/passwordGeneration.service.ts [hunk 3]
```diff
@@ -217,7 +223,82 @@ export class PasswordGenerationService implements PasswordGenerationServiceAbstr
...
+
+        if (policies == null || policies.length === 0) {
+            return enforcedOptions;
+        }
+
+        policies.forEach((currentPolicy) => {
+            if (!currentPolicy.enabled || currentPolicy.data == null) {
+                return;
+            }
+            const currentPolicyData = currentPolicy.data;
```
Comment (kspearrin) [inline src/services/passwordGeneration.service.ts]: Not really necessary. Saving 1 character.

File: src/services/passwordGeneration.service.ts [hunk 4]
```diff
@@ -217,7 +223,82 @@ export class PasswordGenerationService implements PasswordGenerationServiceAbstr
...
+
+            if (currentPolicyData.minLength != null
+                && currentPolicyData.minLength > enforcedOptions.minLength) {
+                enforcedOptions.minLength = currentPolicyData.minLength;
+            }
+
+            if (currentPolicyData.useUpper && !enforcedOptions.useUppercase) {
```
Comment (kspearrin) [inline src/services/passwordGeneration.service.ts]: On these boolean ones, you can skip the `&& !enforcedOptions.useUppercase` again, like we did in the `optionsCache` in the method above. Just set it to true even if it is already true.

File: src/models/domain/passwordGeneratorPolicyOptions.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+export class PasswordGeneratorPolicyOptions extends Domain {
+    minLength: number = 0;
+    useUppercase: boolean = false;
+    useLowercase: boolean = false;
+    useNumbers: boolean = false;
+    numberCount: number = 0;
+    useSpecial: boolean = false;
+    specialCount: number = 0;
+
+    constructor(
```
Comment (kspearrin) [inline src/models/domain/passwordGeneratorPolicyOptions.ts]: Let's just ditch this ctor. Extra code that isn't being used. It's just more to maintain.

File: src/angular/components/password-generator.component.ts (hunk 1 above)
Comment (kspearrin) [reply src/angular/components/password-generator.component.ts]: Separate this from the section above with a new line. Everything else in this "Section" are services.

File: src/services/passwordGeneration.service.ts (hunk 4 above)
Comment (vincentsalucci) [reply src/services/passwordGeneration.service.ts]: Fixed in next update.

File: src/services/passwordGeneration.service.ts (hunk 3 above)
Comment (vincentsalucci) [reply src/services/passwordGeneration.service.ts]: Fixed in next update

File: src/models/domain/passwordGeneratorPolicyOptions.ts (hunk 5 above)
Comment (vincentsalucci) [reply src/models/domain/passwordGeneratorPolicyOptions.ts]: Removed in next update

File: src/services/passwordGeneration.service.ts (hunk 2 above)
Comment (vincentsalucci) [reply src/services/passwordGeneration.service.ts]: Yes - added a few lines to both this function and the `getPasswordGeneratorPolicyOptions` to handle that and keep the UI layer happy at the same time.

**Pull request merged**